src/pairwise.jl

```diff
@@ -1,25 +1,53 @@
-function _pairwise!(::Val{:none}, f, dest::AbstractMatrix, x, y, symmetric::Bool)
-    @inbounds for (i, xi) in enumerate(x), (j, yj) in enumerate(y)
-        symmetric && i > j && continue
-
-        # For performance, diagonal is special-cased
-        if f === cor && eltype(dest) !== Union{} && i == j && xi === yj
-            # TODO: float() will not be needed after JuliaLang/Statistics.jl#61
-            dest[i, j] = float(cor(xi))
-        else
-            dest[i, j] = f(xi, yj)
-        end
+function _pairwise!(::Val{:none}, f, dest::AbstractMatrix{V}, x, y,
+    symmetric::Bool) where {V}
+
+    nr, nc = size(dest)
+
+    # Swap x and y for more efficient threaded loop.
+    if nr < nc
+        dest′ = reshape(dest, size(dest, 2), size(dest, 1))
+        _pairwise!(Val(:none), f, dest′, y, x, symmetric)
+        dest .= transpose(dest′)
+        return dest
     end
-    if symmetric
-        m, n = size(dest)
-        @inbounds for j in 1:n, i in (j+1):m
-            dest[i, j] = dest[j, i]
+
+    #cor and friends are special-cased.
+    iscor = (f in (corkendall, corspearman, cor))
+    (iscor || f == cov) && (symmetric = x === y)
+    #cov(x) is faster than cov(x, x)
+    (f == cov) && (f = ((x, y) -> x === y ? cov(x) : cov(x, y)))
+
+    Threads.@threads for subset in equal_sum_subsets(nr, Threads.nthreads())
+        for j in subset
+            for i = 1:(symmetric ? j : nc)
+                # For performance, diagonal is special-cased
+                if iscor && i==j && y[i] === x[j] && V !== Union{}
+                    dest[j, i] = V === Missing ? missing : 1.0
```

Comment:
I think this optimization might hide errors, e.g. in the case of a 1x1 matrix where the destination has element type `Missing` but `f` would actually return a non-missing value. So I think the only safe optimization might be
```suggestion
                    dest[j, i] = oneunit(f_mn)
```
where `f_mn` is the value of some (probably off-diagonal) `f(x[m], y[n])`.

Reply:
But `iscor` is only `true` for a small set of functions for which we want to set the on-diagonal elements to 1.0, so your scenario "`f` would actually return a non-missing value" cannot apply given that `iscor` is `true`.